sql/src/main/java/io/crate/execution/engine/indexing/ShardDMLExecutor.java

```diff
@@ -155,12 +166,41 @@ private static String getLocalNodeId(ClusterService clusterService) {
         return nodeId;
     }
 
-    private static long processShardResponse(ShardResponse shardResponse) {
+    private static <A> A processResponse(ShardResponse shardResponse, Function<ShardResponse, A> f) {
         Exception failure = shardResponse.failure();
         if (failure != null) {
             Throwables.throwIfUnchecked(failure);
             throw new RuntimeException(failure);
         }
-        return shardResponse.successRowCount();
+        return f.apply(shardResponse);
+    }
+
+    private static Long toRowCount(ShardResponse shardResponse) {
+        return Long.valueOf(processResponse(shardResponse, ShardResponse::successRowCount));
+    }
+
+    private static List<Object[]> toResultRows(ShardResponse shardResponse) {
+        List<Object[]> result = processResponse(shardResponse, ShardResponse::getResultRows);
+        return result == null ? List.of() : result;
     }
+
+    public static final Collector<ShardResponse, long[], Iterable<Row>> ROW_COUNT_COLLECTOR = Collector.of(
+        () -> new long[]{0L},
+        (a, b) -> a[0] += toRowCount(b),
+        (a, b) -> {
+            a[0] += b[0];
+            return a;
+        },
+        a -> List.of(new Row1(a[0]))
+    );
+
+    public static final Collector<ShardResponse, List<Object[]>, Iterable<Row>> RESULT_ROW_COLLECTOR = Collector.of(
+        ArrayList::new,
+        (a, b) -> a.addAll(toResultRows(b)),
+        (a, b) -> {
+            a.addAll(b);
+            return a;
+        },
+        a -> Lists2.map(a, RowN::new)
```

Comment:
I think it would be possible to use `CollectionBucket` here to avoid the copy operation.